sourceinfo/source_code_info.go

```diff
@@ -527,14 +537,121 @@ func (sci *sourceCodeInfo) newLocWithGivenComments(nodeInfo ast.NodeInfo, leadin
 	})
 }
 
+type comments interface {
+	Len() int
+	Index(int) ast.Comment
+}
+
+var emptyComments = subComments{offs: 0, n: 0, c: ast.Comments{}}
+
+type subComments struct {
+	offs, n int
+	c       ast.Comments
+}
+
+func (s subComments) Len() int {
+	return s.n
+}
+
+func (s subComments) Index(i int) ast.Comment {
+	if i < 0 || i >= s.n {
+		panic(fmt.Errorf("runtime error: index out of range [%d] with length %d", i, s.n))
+	}
+	return s.c.Index(i + s.offs)
+}
+
+func (sci *sourceCodeInfo) getLeadingComments(n ast.Node) []comments {
+	s := n.Start()
+	prev := sci.file.PreviousToken(s)
+	info := sci.file.TokenInfo(s)
+	if prev == ast.TokenError {
+		return groupComments(info.LeadingComments())
+	}
+	prevInfo := sci.file.TokenInfo(prev)
+	_, c := attributeComments(prevInfo, info)
+	return c
+}
+
+func (sci *sourceCodeInfo) getTrailingComments(n ast.Node) comments {
+	e := n.End()
+	next := sci.file.NextToken(e)
+	if next == ast.TokenError {
+		return emptyComments
+	}
+	info := sci.file.TokenInfo(e)
+	nextInfo := sci.file.TokenInfo(next)
+	c, _ := attributeComments(info, nextInfo)
+	return c
+}
+
+func attributeComments(prevInfo, info ast.NodeInfo) (t comments, l []comments) {
+	trail := prevInfo.TrailingComments()
+	lead := groupComments(info.LeadingComments())
+
+	if commentsContain(trail, lead, "This is the first detached comment for the syntax.\n") {
+		_ = true
+	}
+
+	if trail.Len() > 0 || len(lead) == 0 {
+		// previous token already has trailing comments or current token
+		// has no comments from which it may borrow
+		return trail, lead
+	}
+	if prevInfo.End().Line == info.Start().Line {
+		// if the tokens are on the same line, no way to attribute
+		return trail, nil
+	}
+	if lead[0].Index(0).Start().Line > prevInfo.End().Line+1 {
+		// first comment is detached from previous token, so can't be a trailing comment
+		return trail, lead
+	}
+	if len(lead) > 1 {
+		// multiple groups? then donate first comment to previous token
+		return lead[0], lead[1:]
+	}
+
+	if lead[0].Index(lead[0].Len()-1).End().Line < info.Start().Line-1 {
+		// there is a blank line between the comments and subsequent token, so
+		// we can donate the comment to previous token
+		return lead[0], lead[1:]
+	}
+	if txt := info.RawText(); len(txt) == 1 && strings.Contains("}])", txt) {
```

Comment:
Previously, [in the lexer](https://github.com/bufbuild/protocompile/pull/46/files#diff-87ceafcdbef726965fe7a7b9e10427cf0c70bfa9b54d2cc1a21530a5d8b5abeaL435), this was an allow-list of punctuation that allowed `".[("`. However, `protoc` actually uses a block-list, to prevent `"}])"`, so I've changed to that, to potentially improve parity between protocompile and `protoc`.